weave/flow/dataset.py

```diff
@@ -236,3 +254,169 @@ def select(self, indices: Iterable[int]) -> Self:
+                }
+
+            # Example 3: Function with default parameters
+            def classify_language(lang, default_category="other"):
+                categories = {"en": "english", "es": "spanish", "fr": "french"}
+                return {"category": categories.get(lang, default_category)}
+
+            # Apply the transformations
+            ds_with_length = ds.map(add_text_length)
+            ds_combined = ds.map(combine_fields)
+            ds_categorized = ds.map(classify_language)
+
+            # Example output for ds_with_length[0]:
+            # {
+            #     "id": 1,
+            #     "text": "Hello world",
+            #     "lang": "en",
+            #     "length": 11
+            # }
+            ```
+        """
+        processed_rows: list[tuple[int, dict]] = []
+
+        # Inspect the function signature
+        sig = inspect.signature(func)
+        param_names = list(sig.parameters.keys())
+
+        # Convert lambda to named function for better async handling
+        func = wrap_lambda(func)
+
+        # Store the function name before applying weave.op
+        func_name = getattr(func, "__name__", "unknown_function")
+
+        # Apply weave.op after lambda conversion
+        func = weave.op(func)
+
+        async def process_row(row: dict) -> dict:
+            try:
+                # Extract and pass specific parameters
+                kwargs = {}
+                for name in param_names:
+                    if name in row:
+                        kwargs[name] = row[name]
+                    else:
+                        # Skip this parameter or use a default if available
+                        if sig.parameters[name].default is not inspect.Parameter.empty:
+                            kwargs[name] = sig.parameters[name].default
+                        else:
+                            raise ValueError(
+                                f"Function expects parameter '{name}' but this column "
+                                f"is not in the dataset row: {short_str(row)}"
+                            )
+                # asyncify the function call
+                result = await async_call(func, **kwargs)
+
+                # Handle the result based on its type
+                if not isinstance(result, dict):
+                    # Raise error for non-dictionary returns
+                    raise TypeError(
+                        f"Function must return a dictionary, but got {type(result).__name__} "
+                        f"from function '{func_name}'. Use a dictionary to specify "
+                        f"which columns to add or update."
+                    )
+            except Exception as e:
+                # Log the error and propagate it to stop the map operation
+                print(f"Error processing row with map function: {short_str(row)}")
+                traceback.print_exc()
+                raise e
+            else:
+                row_copy = copy.deepcopy(row)
```

Comment:
Do you need to do this?  Deepcopy can be expensive for large rows

Reply:
I don't want to modify the underlying dataset, I return a new dataset